src/doc/design_overview/overview.rst

```diff
@@ -0,0 +1,961 @@
+``<Page>`` component (particularly, the connection between
+``active_top_tab`` and the ``<ActiveAppContent>`` sub-component of
+``<Page>``) the ``<ActiveAppContent>`` automatically gets re-rendered, this
+time with the projects page as its contents (which is implemented in
+``smc-webapp/projects.cjsx`` as a React component called
+``<ProjectsPage>``).
+
+Rendering the projects page
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There's something that has been happening in the background all this time
+that we have not talked about yet.  When the client UI was first loaded
+(i.e. ``desktop_app.cjsx``), the sources for several other pages (such
+as the projects page, the account page, etc.) were also loaded.  In
+particular, the projects page is implemented in
+``smc-webapp/projects.cjsx``.  This has some non-trivial side-effects.
+
+One of these side-effects is that the ``ProjectsTable`` class is registered
+with ``AppRedux`` via the ``AppRedux.createTable()`` method.  This class is
+a client-side front-end to the ``"projects"`` table in the DB backend.  This
+creates an instance of a fairly complex object called ``SyncTable`` defined
+in ``smc-util/synctable.coffee``.  We won't go into detail of how this
+works, but basically it keeps a client-side copy of the results of queries
+to the Hub's database, and sets up change listeners that synchronizes those
+cached query results every time the real database changes (via a
+"changefeed", which asynchronously pushes an update to the client every time
+the query result changes--this is a feature around which RethinkDB was
+designed, but this is now being rewritten on top of PostgreSQL).
+
+In effect, while the page has been loading, in the the client app has been
+sending a database query to the server for all the user's projects, and
+receiving the result.  The ``ProjectTable`` has a change handler that
+updates a prop called ``project_map`` which contains the result of the query
+for all the user's projects.  When the ``<ProjectsPage>`` component renders,
+it checks to see if ``project_map`` is undefined.  If it *is* undefined this
+means the database query hasn't completed yet (if the user has no projects
+the result of the query would be an empty list, but ``undefined``).  In this
+case a "Loading..." spinner is rendered.  The React-Redux connection ensures
+that the page is re-rendered whenever ``project_map`` changes, so as soon as
+the database query is completed the page will be re-rendered.
+
+The rest is fairly straightforward, given an understanding of React.  The
+``<ProjectsPage>`` component consists of a number of sub-components,
+including one called ``<ProjectList>`` where the all the projects are looped
+over and displayed in a table.  There are some additional complications
+related to the ability to filter projects in various ways, but there's
+nothing special to this.
+
+Conclusion
+^^^^^^^^^^
+
+And that's it! We got from an empty browser window the user's projects
```

Comment:
Yes, I just started looking at that last Thursday (e.g. syncdoc.coffee).  I think this example was meant to be very simple, but it would be interesting when I have more time to write up a more complicated example involving, say, the Sage worksheet.  I think it's very clever how it's built on top of codemirror--that surprised me a lot.  There are some things about it that I think are a bit clunkier UI-wise than the Jupyter Notebook, but I can also see the advantages to this approach.  I think I could learn to like it if it just had a vi mode :)